clang/include/clang/Basic/DiagnosticFrontendKinds.td

```diff
@@ -330,6 +330,14 @@ def warn_alias_with_section : Warning<
   "as the %select{aliasee|resolver}2">,
   InGroup<IgnoredAttributes>;
 
+let CategoryName = "Bounds Safety Issue" in {
+def err_bounds_safety_lang_not_supported : Error<
+  "bounds safety is only supported for C">;
+def warn_bounds_safety_asm_ignored : Warning<
+  "'-fbounds-safety' is ignored for assembly">,
```

Comment:
Nit. In this PR the flag is `-fbounds-safety-experimental` not `-fbounds-safety`.

Reply:
For assembly input, we generally ignore handling the option and leave a warning https://maskray.me/blog/2023-08-25-clang-wunused-command-line-argument#assembly-files

```
% fclang -fsanitize=address -fpatchable-function-entry=2 -c a.s
clang: warning: argument unused during compilation: '-fsanitize=address' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-fpatchable-function-entry=2' [-Wunused-command-line-argument]
```

Reply:
@MaskRay Thank you! I can use `warning: argument unused during compilation` for the driver. Do you know what is the convention for unused flags in cc1?

Reply:
https://maskray.me/blog/2023-08-25-clang-wunused-command-line-argument has some notes. If you don't call functions like `getLastArg` that `claim` the option, there will be a `warning: argument unused during compilation: '-fsanitize=address' [-Wunused-command-line-argument]` for assembly input.

Reply:
@MaskRay So we aren't explicit using `getLastArg` but it looks like we are calling it indirectly from `addOptInFlag(...)` which I think means we are unconditionally claiming the option.  In order to make your suggestion work we'd have to conditionally call `addOptInFlag` in the driver based on the selected input language but I don't think that's going to work because no `LangOptions` object exists for us query at that point in the driver AFAICT.

Also future patches we will upstream will add additional `Args.getLastArg(options::OPT_fbounds_safety_experimental, ...)` calls in the driver which means the option will get claimed there too which will break any reliance on the option not being claimed that might be implemented in this PR.

This warning is really a frontend warning and not a driver warning so I think we should use the frontend warning that @rapidsna introduced in this PR.

Reply:
Conventionally the language compatibility checking and other checking is performed in Driver, not in Frontend.. If you move the language check to Driver, the diagnostic will be natural since clang integrated assembler uses `ClangAs` instead of `Clang`.

Reply:
@MaskRay Thanks. I still need the error for `-fbounds-safety` not being supported in C++/Obj-C/etc in the frontend because it needs to be an error for `cc1`.

I will change it so that the unused warning for assembly to be handled in the driver like the rest of the unused options for assembly.

Reply:
> Conventionally the language compatibility checking and other checking is performed in Driver, not in Frontend.. If you move the language check to Driver, the diagnostic will be natural since clang integrated assembler uses `ClangAs` instead of `Clang`.

@MaskRay It seems both `Clang` and `ClangAs` are invoked for the `clang` command with an assembly input (i.e., `ClangAs` is invoked after `Clang`) so it seems to me that most options are already `claimed` in `Clang` and so `warning: argument unused during compilation` doesn't seem to fire for `clang` in most cases including `-fsanitize=address`.

Also, we need to have the input language (`InputKind`) to report the right diagnostics so the frontend still seems like the right place to handle this.

Reply:
@MaskRay Sorry, it seems we can still do it in Driver with `clang::driver::types`. I'm going to make a change to report the "unused" warning from the driver, and report the "unsupported language" error in the frontend. Does it sound okay to you?

Reply:
> @MaskRay It seems both `Clang` and `ClangAs` are invoked for the `clang` command with an assembly input (i.e., `ClangAs` is invoked after `Clang`) so it seems to me that most options are already `claimed` in `Clang` and so `warning: argument unused during compilation` doesn't seem to fire for `clang` in most cases including `-fsanitize=address`.

I just confirmed that this is happening when the input file format is `.s` and no `-x` is provided from the command line. If we do `-x assembler` then only `ClangAs` is launched and the unused option warning fires as expected.

Reply:
@MaskRay I removed the frontend warning, so that it follows how the driver handles unused flags. Now only `warning: argument unused during compilation` fires for `-x assembler`, and not for `-x assembler-with-cpp` to conform to most of other options.